src/lib/party.ts

```diff
@@ -523,8 +571,94 @@ class Party {
     return clonedParty;
   }
 
+  token() {
+    let thisParty = this;
+    let customToken = new Token({
+      tokenOwner: thisParty.body.publicKey,
+      parentTokenId: thisParty.body.tokenId,
+    });
+
+    return {
+      id: customToken.id,
+      parentTokenId: customToken.parentTokenId,
+      tokenOwner: customToken.tokenOwner,
+
+      mint({ address, amount }: MintOrBurnParams) {
+        let receiverParty = createChildParty(thisParty, address, {
+          caller: this.id,
+          tokenId: this.id,
+        });
+
+        // Add the amount to mint to the receiver's account
+        receiverParty.body.balanceChange =
+          receiverParty.body.balanceChange.add(amount);
+      },
+
+      burn({ address, amount }: MintOrBurnParams) {
+        let receiverParty = createChildParty(thisParty, address, {
+          caller: this.id,
+          tokenId: this.id,
+          useFullCommitment: Bool(true),
+        });
+
+        // Sub the amount to burn from the receiver's account
+        receiverParty.body.balanceChange =
+          receiverParty.body.balanceChange.sub(amount);
+
+        // Require signature from the receiver account being deducted
```

Comment:
Is this always true? Can't certain tokens be burned via proof?

Reply:
Good question, I'm not sure how this would work with the current SnarkyJS API. @mitschabaude, is there a way to require either a proof or signature? Also, how would one construct a proof to do this?

Reply:
Aaah yes. I forgot about that. It depends on the senderParty's account, and the permissions on there, whether it needs a proof or a signature.

Reply:
This is a puzzle 🤔 because different zkApps have completely different ways of authorizing transactions. So my fear is token contracts will typically impose a certain interface on zkApps for that token.

I'm pretty sure @mrmr1993 has thought deeper about this

In any case, I think there's no easy solution here. The `token()` API we designed seems fundamentally not flexible enough to handle parties with proofs, because they are all different. The API must be one where users provide their own parties, not where we create parties for them. But even then it seems very tricky to make a circuit for the token owner contract that's flexible enough to handle arbitrary zkApp children. So I'm not even sure we can _ever_ provide a ready-made, general implementation for parties with proofs.

That makes it all the more important to at least expose the low-level pieces to create token contracts -- I think that's the `Token` class (which I still think should be just a function)

Reply:
Maybe a fairly general solution is not that hard to build after all.
The token() functions must witness their child parties instead of hardcoding them into the circuit, and only assert the properties of them that they care about -  balanceChanges and tokenId.

The challenge is how to create a good API for this. I imagine its functions must take a party as input, which might mean we have to make Party one of the possible arguments of a smart contract method. A smart contract could sort of "call" a token contract with 'this' as an argument, and that token contract would make the smart contract its child. A short of reversed zkapp composability

Reply:
I guess the natural API from the protocol perspective would be if the token contract methods would accept a callback, which is like a transaction block in that you can run a (or possibly multiple) smart contract methods in it, which would create parties that are authorized with a proof. And the token contract would then assert certain conditions on the resulting parties, for example
* They can't have children
* There must be exactly / at most X of them
* The balances must cancel in send
* In mint / burn, conditions X, Y, Z must hold

Reply:
I created a separate issue so that we can discuss the priority & path forward on this. It's out of scope for this PR, it's already nice to have token transfers authorized by a proof
https://github.com/o1-labs/snarkyjs/issues/300